### PR TITLE
[sycl-post-link] Split SYCL and ESIMD kernels into separate modules

### PR DESCRIPTION
**Pull request by DenisBakhvalov**
For now, this change doesn't have any effect on existing programs since we don't allow mixing SYCL and ESIMD kernels in one source or in one program.
But this is an essential step towards allowing such a mix since ESIMD kernels require specific processing as opposed to usual SYCL kernels.

PLEASE, only review commit: 147eb5fa54a9eb5f5a376bb656911396d873d59a

Patch (1/3): https://github.com/intel/llvm/pull/3042
Patch (2/3): https://github.com/intel/llvm/pull/3043
Patch (3/3): this PR

### CODE REVIEW

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:460]: Am I right that this filename change was only introduced to simplify debugging, i.e. there is no functional need for it?

I have a small concern that a number of functions were changed just to have this prefix added and it is kind of a non-generic solution (like what if we decided to add one more level of device code split for some other feature?)

Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:460]: Before my change the code was split into modules and saved to files with the name derived from `OutputFilename` plus sequential index, i.e. `/path/to/file_01.bc`, `/path/to/file_02.bc`, etc. 

Now, when we split ESIMD modules we cannot overwrite already processed SYCL files, this is why we need to change the name of the ESIMD files. We could have:
```
`/path/to/file_01.bc` // sycl
`/path/to/file_02.bc` // sycl
`/path/to/file_esimd_01.bc` // esimd
`/path/to/file_esimd_02.bc` // esimd
```
or
```
`/path/to/file_01.bc` // sycl
`/path/to/file_02.bc` // sycl
`/path/to/file_03.bc` // esimd
`/path/to/file_04.bc` // esimd
```
I choose the first scheme, since, as you said, it will be easier to find ESIMD modules, and the names of the files do not affect functionality. But I'm OK to change if you think the second is better. However, I think we need to propagate the `IsEsimd` flag (or other parameter) regardless of the naming scheme.

We can introduce a class that will keep the state. Then all the propagation will go away. But we can do that later in further patches.

Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:460]: > I choose the first scheme, since, as you said, it will be easier to find ESIMD modules, and the names of the files do not affect functionality. But I'm OK to change if you think the second is better. 

I don't have a strong opinion here, so, the proposed way looks fine to me. We can hear from @kbobrovs here to have another opinion

> However, I think we need to propagate the IsEsimd flag (or other parameter) regardless of the naming scheme.

Right, but from functional point of view it is only needed in `processOneModule`

Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:460]: > We can introduce a class that will keep the state. Then all the propagation will go away. But we can do that later in further patches.

Personally, I'm not a fun of such states, because it makes it harder to understand where particular value is came from and how the call graph looks like

**Review by AlexeySachkov — APPROVED** [review]
This change should probably be updated to ToT because of the conflict, but in general I don't have any objections, LGTM

File: llvm/test/tools/sycl-post-link/sycl-esimd/basic-sycl-esimd-split.ll
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/sycl-esimd/basic-sycl-esimd-split.ll:42]: I expect `_Z28__spirv_GlobalInvocationId_xv` to be part of both modules (cloned) - can you please add a check for that?

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll:60]: please add call graph nodes into CHECK-*s as well

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-source.ll
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-source.ll:74]: please add call graph nodes into CHECK-*s as well

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll:54]: please add call graph nodes into CHECK-*s as well
also this test could probably be merged with the previous one as -symbols and split type are "orthogonal"

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:133]: general comment: there will be no way to tell the post link not to perform SYCL/ESIMD splitting. It might not be always necessary (and ideally, GPU BEs will accept both kinds of kernels in one module in future), so I'd suggest to add something like `-no_sycl_esimd_split` to preserve previous behavior.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:460]: Current scheme with `_esimd` suffix seems clearer to me. Users don't care, but devs will appreciate it.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:449]: I suggest to use `StringRef suffix` instead of `bool IsEsimd`, as filename generation uses this parameter only to create suffix. This would reduce 'IsEsimd' knowledge spread

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:723]: I think SYCL_EXTERNAL functions are missing. They are also "roots" of the modules. See also comment about SYCL_EXTERNAL above. We might not be supporting them properly in sycl-post-link at all.

On the other hand, post-link happens to be invoked on already linked code, so this is not high-priority. A TODO + issue would be enough in that case, I think

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll
Comment (kbobrovs) [inline llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll:18]: What is behavior on SYCL_EXTERNAL functions? I understand this is not ESIMD-specific, but if there are gaps then it would be good to file an issue on that at least. This affects ESIMD as well.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:859]: it seems worthwhile to issue a warning to let user know that ESIMD part if ignored.

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll
Comment (DenisBakhvalov) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll:54]: My idea was that SYCL-ESIMD splitting always happens even without -split option. I think it is more clear that way. Those tests are blazingly fast, so I think there is no problem having it as a separate test.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:725]: I'm not sure I understand what would be the scenario with SYCL_EXTERNAL functions and what we should do with them.

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll
Comment (kbobrovs) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll:54]: > Those tests are blazingly fast, so I think there is no problem having it as a separate test.

OK, agreed.

> My idea was that SYCL-ESIMD splitting always happens even without -split option.

After thinking a bit more on this, such behavior of sycl-post-link would seem unexpected - it does something when not asked. My previous suggestion was to add `-no-sycl-esimd-split`, but it seems `-split-esimd` would be preferable, which means not doing esimd split w/o this option. How do you think? Adding @bader, @AlexeySachkov as well.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:725]: The scenario is when sycl-post-link is run on a static library with SYCL_EXTERNAL functions of both kinds - ESIMD and SYCL.

**Comment by bader** [comment]
@DenisBakhvalov, please, resolve merge conflicts.

**Comment by DenisBakhvalov** [comment]
> @DenisBakhvalov, please, resolve merge conflicts.

Done.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:890]: I introduced a new `-split-esimd` option, which is not compatible with `-ir-output-only`.

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll
Comment (DenisBakhvalov) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll:18]: We discussed it and decided to leave it as it is for now.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:727]: We discussed it and decided to leave it as it is for now.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Done.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:137]: I introduced a new `-split-esimd` option that controls SYCL-ESIMD splitting.

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll
Comment (DenisBakhvalov) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-symbols.ll:54]: I introduced a new `-split-esimd` option that controls SYCL-ESIMD splitting.

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-source.ll
Comment (DenisBakhvalov) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-source.ll]: Done

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll
Comment (DenisBakhvalov) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll]: Done

File: llvm/test/tools/sycl-post-link/sycl-esimd/basic-sycl-esimd-split.ll
Comment (DenisBakhvalov) [reply llvm/test/tools/sycl-post-link/sycl-esimd/basic-sycl-esimd-split.ll]: Done

**Comment by DenisBakhvalov** [comment]
@kbobrovs, thanks for the review. I fixed your comments in 0b6c9c7813c9350ee5308a65cd12bdaad11c1f0b. Please review.

File: llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll
Comment (kbobrovs) [reply llvm/test/tools/sycl-post-link/sycl-esimd/sycl-esimd-split-per-kernel.ll:18]: right, sounds good.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:727]: right, seems a very-low priority topic, as post-link is to be invoked on already linked code

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:798]: Suggest to add:
Functions unreachable from any kernel (including those marked as with "explicit SIMD" metadata) are dropped from the resulting module(s).

Comment (DenisBakhvalov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Added

**Review by kbobrovs — APPROVED** [review]
LGTM

**Comment by kbobrovs** [comment]
> This change should probably be updated to ToT because of the conflict, but in general I don't have any objections, LGTM

treating this as approval from @AlexeySachkov. No other objections - merging...

**Pull request merged**
